### PR TITLE
Jormungandr: create a connector for Instant System

### PR DESCRIPTION
**Pull request by xlqian**
https://jira.canaltp.fr/browse/NAVP-714

- [x] add tests

### CODE REVIEW

File: source/jormungandr/jormungandr/street_network/instance_system.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+# Stay tuned using
+# twitter @navitia
+# IRC #navitia on freenode
+# https://groups.google.com/d/forum/navitia
+# www.navitia.io
+
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/street_network/instance_system.py]: You forgot the mandatory import from future to make thing look a bit more python 3.

File: source/jormungandr/jormungandr/street_network/instance_system.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+from ride_sharing_journey import RideSharingJourney
+
+
+def _record_external_failure(message, sn_system_id):
+    utils.record_external_failure(message, 'streetnetwork', sn_system_id)
+
+
+def _call_service(url, api_key, sn_system_id):
+    logger = logging.getLogger(__name__)
+    breaker = pybreaker.CircuitBreaker(fail_max=100, reset_timeout=1000)
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/street_network/instance_system.py]: The breaker is useless if it isn't persistent between call

File: source/jormungandr/jormungandr/street_network/instance_system.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    headers = {u'Authorization': u'apiKey {}'.format(api_key)}
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/street_network/instance_system.py]: Once `unicode_literals` imported you won"t need the `u` anymore

File: source/jormungandr/jormungandr/street_network/instance_system.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    try:
+        return breaker.call(requests.get, url=url, headers=headers, timeout=1000)
+    except pybreaker.CircuitBreakerError as e:
+        logger.error('Ride sharing service dead (error: {})'.format(e))
+        _record_external_failure('circuit breaker open', sn_system_id)
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/street_network/instance_system.py]: It's better to use exception to handle that like it has been done in the realtime schedules part. 
But we will do the refactoring latter.

File: source/jormungandr/jormungandr/street_network/instance_system.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+from jormungandr import utils
+from ride_sharing_journey import RideSharingJourney
+
+
+def _record_external_failure(message, sn_system_id):
+    utils.record_external_failure(message, 'streetnetwork', sn_system_id)
+
+
+def _call_service(url, api_key, sn_system_id):
+    logger = logging.getLogger(__name__)
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/street_network/instance_system.py]: Always add the `sn_system_id` as metadata to the logger

File: source/jormungandr/jormungandr/street_network/instance_system.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        _record_external_failure(str(e), sn_system_id)
+    return None
+
+
+def _make_timestamp_from_str(strftime):
+    return time.mktime(parser.parse(strftime).utctimetuple())
+
+
+def _get_ridesharing_journeys(raw_journeys):
+    def has_ridesharing_path(j):
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/street_network/instance_system.py]: This handle ours journeys or their?

File: source/jormungandr/jormungandr/street_network/instance_system.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+            res.total_seats = None
+            res.available_seats = p['rideSharingAd']['vehicle']['availableSeats']
+
+            ridesharing_journeys.append(res)
+
+    return ridesharing_journeys
+
+
+def request_system_instance(api_key, from_coord, to_coord, request_datetime, datetime_represent=True, limit=-1):
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/street_network/instance_system.py]: It's probably a typo (like in your filename) as it should be `instant_system`

File: source/jormungandr/jormungandr/street_network/ride_sharing_journey.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+# Stay tuned using
+# twitter @navitia
+# IRC #navitia on freenode
+# https://groups.google.com/d/forum/navitia
+# www.navitia.io
+
+
+class RideSharingJourney(object):
+    # https://stackoverflow.com/a/28059785/1614576
+    __slots__ = ('duration',
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/street_network/ride_sharing_journey.py]: Probably a bit over engineered here but :+1:

**Review by kinnou02 — COMMENTED** [review]
You regulary made a typo on "Instant System".
Also, since this connector isn't comaptible with other street_network connector we should put it elsewhere.

Not using a class will probably lead to code duplication with error handling and logging setup too.

File: source/jormungandr/jormungandr/street_network/ride_sharing_journey.py (hunk 8 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/street_network/ride_sharing_journey.py]: i agree, but this technique may be very useful in kirin :)

File: source/jormungandr/jormungandr/street_network/ride_sharing_journey.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+                 'pickup_lat',
+                 'pickup_lon',
+                 'dropoff_addr',
+                 'dropoff_lat',
+                 'dropoff_lon',
+                 'pickup_date_time',
+                 'dropoff_date_time',
+                 'driver_name',
+                 'driver_gender',
+                 'dirve_image_url',
```
Comment (pbougue) [inline source/jormungandr/jormungandr/street_network/ride_sharing_journey.py]: typo: `dirve`

File: source/jormungandr/jormungandr/street_network/ride_sharing_journey.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+                 'offer_url',
+                 'pickup_addr',
+                 'pickup_lat',
+                 'pickup_lon',
+                 'dropoff_addr',
+                 'dropoff_lat',
+                 'dropoff_lon',
+                 'pickup_date_time',
+                 'dropoff_date_time',
+                 'driver_name',
```
Comment (pbougue) [inline source/jormungandr/jormungandr/street_network/ride_sharing_journey.py]: A "individual" structure might be nice, to mutualize with potential passengers in the future

File: source/jormungandr/jormungandr/street_network/ride_sharing_journey.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+class RideSharingJourney(object):
+    # https://stackoverflow.com/a/28059785/1614576
+    __slots__ = ('duration',
+                 'distance',
+                 'shape',
+                 'offer_url',
+                 'pickup_addr',
+                 'pickup_lat',
+                 'pickup_lon',
+                 'dropoff_addr',
```
Comment (pbougue) [inline source/jormungandr/jormungandr/street_network/ride_sharing_journey.py]: this "place" object can be mutualized

File: source/jormungandr/jormungandr/street_network/ride_sharing_journey.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+# https://groups.google.com/d/forum/navitia
+# www.navitia.io
+
+
+class RideSharingJourney(object):
+    # https://stackoverflow.com/a/28059785/1614576
+    __slots__ = ('duration',
+                 'distance',
+                 'shape',
+                 'offer_url',
```
Comment (pbougue) [inline source/jormungandr/jormungandr/street_network/ride_sharing_journey.py]: I called it "ridesharing_ad" in RFC, maybe also in proto : we have to adjust naming a bit

File: source/jormungandr/jormungandr/street_network/instance_system.py (hunk 6 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/street_network/instance_system.py]: theirs, it gives us pure ridesharing journeys

File: source/jormungandr/jormungandr/scenarios/ridesharing/instant_system.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+                                                reset_timeout=app.config['CIRCUIT_BREAKER_INSTANT_SYSTEM_TIMEOUT_S'])
+
+    def _call_service(self, params):
+        self.logger.debug("requesting instant system")
+
+        headers = {'Authorization': 'apiKey {}'.format(self.api_key)}
+        try:
+            return self.breaker.call(requests.get, url=self.api_url, headers=headers, params=params, timeout=1000)
+        except pybreaker.CircuitBreakerError as e:
+            self.logger.error('Ride sharing service dead (error: {})'.format(e))
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/ridesharing/instant_system.py]: very minor: we use one same word 'ridesharing', and I'd mention `Instant System`

File: source/jormungandr/jormungandr/scenarios/ridesharing/instant_system.py [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+
+        return ridesharing_journeys
+
+    def request_instant_system(self, from_coord, to_coord, request_datetime, datetime_represent, limit=None):
+        """
+
+        :param from_coord: lat,lon ex: '48.109377,-1.682103'
+        :param to_coord: lat,lon ex: '48.020335,-1.743929'
+        :param request_datetime: a timestamp (utc)
+        :param datetime_represent: bool
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/ridesharing/instant_system.py]: as it's a bool `datetime_represents_start` would be way clearer
Or use a `utils.PeriodExtremity`

File: source/jormungandr/jormungandr/scenarios/ridesharing/instant_system.py [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+        :param limit: optional
+        :return:
+        """
+        # TODO: url and apiKey should be read from config
+        # format of datetime: 2017-12-25T07:00:00Z
+        datetime_str = datetime.datetime.fromtimestamp(request_datetime).strftime('%Y-%m-%dT%H:%M:%SZ')
+
+        params = {'from': from_coord,
+                  'to': to_coord,
+                  ('arrivalDate', 'departureDate')[datetime_represent]: datetime_str}
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/ridesharing/instant_system.py]: not super fan of the safety of this hack :)

File: source/jormungandr/jormungandr/scenarios/ridesharing/tests/instant_system_tests.py [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+
+        ridesharing_journeys = instant_system.request_instant_system(from_coord=from_coord,
+                                                                     to_coord=to_coord,
+                                                                     datetime_represent=datetime_represent,
+                                                                     request_datetime=request_datetime)
+
+        assert len(ridesharing_journeys) == 2
+        assert ridesharing_journeys[0].pickup_place.addr == "9 Allée Rochester, Rennes"
+        assert ridesharing_journeys[0].dropoff_place.addr == "2 Avenue Alphonse Legault, Bruz"
+        assert ridesharing_journeys[0].pickup_date_time == 1514182079
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/ridesharing/tests/instant_system_tests.py]: I think you can use utils tools to have a clearer test, maybe not the same one you use in code :)

**Review by pbougue — COMMENTED** [review]
Quite minor comments (can be corrected in my PR after merge)

File: source/jormungandr/jormungandr/scenarios/ridesharing/tests/instant_system_tests.py [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+        period_extremity = utils.PeriodExtremity(datetime=1514181600,
+                                                represents_start=True)
+        ridesharing_journeys = instant_system.request_instant_system(from_coord=from_coord,
+                                                                     to_coord=to_coord,
+                                                                     period_extremity=period_extremity)
+
+        assert len(ridesharing_journeys) == 2
+        assert ridesharing_journeys[0].pickup_place.addr == "9 Allée Rochester, Rennes"
+        assert ridesharing_journeys[0].dropoff_place.addr == "2 Avenue Alphonse Legault, Bruz"
+        assert ridesharing_journeys[0].pickup_date_time == utils.str_to_time_stamp("20171225T060759")
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/ridesharing/tests/instant_system_tests.py]: Those timestamps from a `2017-12-25T08:07:59+01:00` looks really weird... You probably have a problem.

**Review by pbougue — APPROVED** [review]

**Comment by pbougue** [comment]
Jenkins, test this please

**Comment by xlqian** [comment]
Jenkins, test this please!

**Pull request merged**